sdks/python/apache_beam/io/filebasedsource.py

```diff
@@ -449,3 +458,92 @@ def expand(self, pvalue):
+               desired_bundle_size,  # type: int
+               min_bundle_size,  # type: int
+               source_from_file,  # type: Callable[[str], iobase.BoundedSource]
+               with_filename=False,  # type: bool
+               **kwargs  # parameters for MatchContinuously
+              ):
+    """
+    Args:
+      file_pattern: a file pattern to match
+      splittable: If False, files won't be split into sub-ranges. If True,
+                  files may or may not be split into data ranges.
+      compression_type: A ``CompressionType`` object that specifies the
+                  compression type of the files that will be processed. If
+                  ``CompressionType.AUTO``, system will try to automatically
+                  determine the compression type based on the extension of
+                  files.
+      desired_bundle_size: the desired size of data ranges that should be
+                           generated when splitting a file into data ranges.
+      min_bundle_size: minimum size of data ranges that should be generated when
+                           splitting a file into data ranges.
+      source_from_file: a function that produces a ``BoundedSource`` given a
+                        file name. System will use this function to generate
+                        ``BoundedSource`` objects for file paths. Note that file
+                        paths passed to this will be for individual files, not
+                        for file patterns even if the ``PCollection`` of files
+                        processed by the transform consist of file patterns.
+      with_filename: If True, returns a Key Value with the key being the file
+        name and the value being the actual data. If False, it only returns
+        the data.
+
+    refer to ``MatchContinuously`` for additional args including 'interval',
+    'has_deduplication', 'start_timestamp', 'stop_timestamp',
+    'match_updated_files'.
+    """
+    self._file_pattern = file_pattern
+    self._splittable = splittable
+    self._compression_type = compression_type
+    self._desired_bundle_size = desired_bundle_size
+    self._min_bundle_size = min_bundle_size
+    self._source_from_file = source_from_file
+    self._with_filename = with_filename
+    self._kwargs_for_match = {
+        k: v
+        for (k, v) in kwargs.items() if k in self.ARGS_FOR_MATCH
+    }
+
+  def expand(self, pbegin):
+    # imported locally to avoid circular import
+    from apache_beam.io.fileio import MatchContinuously
+
+    return (
+        pbegin
+        | MatchContinuously(self._file_pattern, **self._kwargs_for_match)
+        | 'ExpandIntoRanges' >> ParDo(
+            _ExpandIntoRanges(
+                self._splittable,
+                self._compression_type,
+                self._desired_bundle_size,
+                self._min_bundle_size,
+                do_match=False))
+        # | 'Reshard' >> Reshuffle() # not Reshuffle because needs timely read.
```

Comment:
The side effect of reshuffle makes it pretty hard to ensure _ReadRange is executed instantly after matchAll. group by key will hold elements until they get fired. I tried adding windows and triggers but it still did not work well, and makes the pipeline complicated.

The question is, what is the consequence if we do not reshuffle here? Does it cause wrong data (emit more than once) or data losing if one read fails, or just makes the worker do unnecessary retry? If it is just the later I think it is fine here.

Reply:
Since due to that triggers for fn runner is WIP. Raised [BEAM-14497](https://issues.apache.org/jira/browse/BEAM-14497).